server/debugsession.go

```diff
@@ -0,0 +1,834 @@
+		address := sdk.Address(addressLocation.Address)
+		// nolint:staticcheck
+		account, err = backendEmulator.GetAccountUnsafe(address)
+		if err != nil {
+			return ""
+		}
+
+		contract, ok := account.Contracts[addressLocation.Name]
+		if !ok {
+			return ""
+		}
+
+		return string(contract)
+	}
+
+	return ""
+}
+
+func (ds *debugSession) step() {
+	ds.debugger.RequestPause()
+	ds.debugger.Continue()
+}
+
+func (ds *debugSession) run() context.CancelFunc {
+	ds.backend.GetEmulator().SetDebugger(ds.debugger)
+	if ds.stopOnEntry {
+		ds.debugger.RequestPause()
+	}
+	ds.variableHandleCounter = 0
+	ds.variables = make(map[int]any, 0)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+
+			case stop := <-ds.debugger.Stops():
+				ds.stop = &stop
+				depth := len(ds.stop.Interpreter.CallStack())
+
+				//TODO: check stop reason breakpoint
+
+				if ds.targetDepth == -1 || depth <= ds.targetDepth {
+					ds.send(&dap.StoppedEvent{
+						Event: newDAPEvent("stopped"),
+						Body: dap.StoppedEventBody{
+							Reason:            "pause",
+							AllThreadsStopped: true,
+							ThreadId:          1,
+						},
+					})
+				} else {
+					ds.step()
+				}
+
+			}
+		}
+	}()
+
+	if ds.code != "" {
+		go func() {
+			// TODO: add support for arguments
+			// TODO: add support for transactions. requires automine
+
+			result, err := ds.backend.ExecuteScriptAtLatestBlock(context.Background(), []byte(ds.code), nil)
+			cancel()
+
+			var outputBody dap.OutputEventBody
+			if err != nil {
+				outputBody = dap.OutputEventBody{
+					Category: "stderr",
+					Output:   err.Error(),
+				}
+			} else {
+				outputBody = dap.OutputEventBody{
+					Category: "stdout",
+					Output:   string(result),
+				}
+			}
+
+			ds.send(&dap.OutputEvent{
+				Event: newDAPEvent("output"),
+				Body:  outputBody,
+			})
+
+			var exitCode int
+			if err != nil {
+				exitCode = 1
+			}
+
+			ds.send(&dap.ExitedEvent{
+				Event: newDAPEvent("exited"),
+				Body: dap.ExitedEventBody{
+					ExitCode: exitCode,
+				},
+			})
+
+			ds.send(&dap.TerminatedEvent{
+				Event: newDAPEvent("terminated"),
+			})
+
+			ds.backend.GetEmulator().EndDebugging()
+
+		}()
+
+	}
+	ds.launched = true
+	return cancel
+}
+
+func newDAPEvent(event string) dap.Event {
+	return dap.Event{
+		ProtocolMessage: dap.ProtocolMessage{
+			Seq:  0,
+			Type: "event",
+		},
+		Event: event,
+	}
+}
+
+func newDAPResponse(requestSeq int, command string, success bool) dap.Response {
+	return dap.Response{
+		ProtocolMessage: dap.ProtocolMessage{
+			Seq:  0,
+			Type: "response",
+		},
+		Command:    command,
+		RequestSeq: requestSeq,
+		Success:    success,
+	}
+}
+
+func newDAPSuccessResponse(request *dap.Request) dap.Response {
+	return newDAPResponse(request.Seq, request.Command, true)
+}
+
+func newDAPErrorResponse(requestSeq int, command string, message dap.ErrorMessage) *dap.ErrorResponse {
```

Comment:
nit:
```suggestion
func newDAPErrorResponse(request *dap.Request, message dap.ErrorMessage) *dap.ErrorResponse {
```

Reply:
Refactored in 87c3211102178900f8925d4877b0a60f9ff986bb